services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/NewSemanticUsageValidator.java

```diff
@@ -0,0 +1,253 @@
+    return SEMANTIC_VALIDATOR;
+  }
+
+  /**
+   * General method to check if the value is non-default for the given field.
+   * @param object the value to check
+   * @param currentField the field to check
+   * @param targetField the target field to check
+   * @return true if the value is non-default, false otherwise
+   */
+  public boolean isNonDefaultValue(Object object, Schema.Field currentField, Schema.Field targetField) {
+    switch (currentField.schema().getType()) {
+      case UNION:
+        return isNonDefaultValueUnion(object, currentField);
+      case ENUM:
+        return isNewEnumValue(object, currentField, targetField);
+      case FIXED:
+        if (currentField.schema().getFixedSize() != targetField.schema().getFixedSize()) {
+          return returnTrueAndLogError(
+              String.format("Field %s: has different fixed size", formatFieldName(currentField.name())));
+        }
+        return false;
+      default:
+        return isNonDefaultValueField(object, currentField);
+    }
+  }
+
+  /**
+   * Check if the value is a new enum value.
+   * @param object the value to check
+   * @param currentField the field to check with enum type and its current symbols
+   * @param targetField the field to check with enum type and its target symbols
+   * @return true if the value is a new enum value, false otherwise
+   */
+  public boolean isNewEnumValue(Object object, Schema.Field currentField, Schema.Field targetField) {
+    List<String> currentSymbols = currentField.schema().getEnumSymbols();
+    List<String> targetSymbols = targetField.schema().getEnumSymbols();
+    if (targetSymbols.stream().anyMatch(symbol -> !currentSymbols.contains(symbol) && symbol.equals(object))) {
+      return returnTrueAndLogError(
+          String.format("Field %s contains new enum value: %s", formatFieldName(currentField.name()), object));
+    }
+    return false;
+  }
+
+  /**
+   * Check if the value is non-default for the given union field.
+   * If the value is null, it is considered default.
+   * If object is non-null, we need to find the schema for the object inside the union and check the default value.
+   * @param object the value to check
+   * @param currentField the field to check with union type
+   * @return true if the value is non-default, false otherwise
+   */
+  public boolean isNonDefaultValueUnion(Object object, Schema.Field currentField) {
+    if (object == null) {
+      return false;
+    }
+
+    Schema subSchema = findObjectSchemaInsideUnion(object, currentField);
+    if (subSchema == null) {
+      return returnTrueAndLogError(
+          String.format(
+              "Field %s: Cannot find the match schema for value %s from schema union",
+              formatFieldName(currentField.name()),
+              object));
+    }
+
+    Schema.Field newCurrentField = AvroCompatibilityHelper.createSchemaField(currentField.name(), subSchema, "", null);
```

Comment:
Here we are using "null" as default value; if subSchema is int or double, "null" would be invalid default value.